lib/dynamoid/persistence.rb

```diff
@@ -216,7 +216,21 @@ def destroy
     # @since 0.2.0
     def delete
       options = range_key ? {:range_key => dump_field(self.read_attribute(range_key), self.class.attributes[range_key])} : {}
+
+      # Add an optimistic locking check if the lock_version column exists
+      if(self.class.attributes[:lock_version])
+        conditions = {:if => {}}
+        conditions[:if][:lock_version] =
+          if changes[:lock_version].nil?
+            self.lock_version
+          else
+            changes[:lock_version][0]
+          end
+        options[:conditions] = conditions
+      end
       Dynamoid.adapter.delete(self.class.table_name, self.hash_key, options)
+    rescue Dynamoid::Errors::ConditionalCheckFailedException
+      raise Dynamoid::Errors::StaleObjectError.new(self, 'delete')
```

Comment:
Sure, would you like me to convert it to string for attempted_action attribute on the exception. It would be a breaking change otherwise.